smac/intensification/successive_halving.py

```diff
@@ -567,70 +578,144 @@ def _compare_configs(self,
-            chall_runs = run_history.get_runs_for_config(challenger, only_max_observed_budget=True)
-            if len(inc_runs) > 1:
-                raise ValueError(
-                    'Number of incumbent runs on budget %f must not exceed 1, but is %d',
-                    inc_runs[0].budget, len(inc_runs),
-                )
-            if len(chall_runs) > 1:
-                raise ValueError(
-                    'Number of challenger runs on budget %f must not exceed 1, but is %d',
-                    chall_runs[0].budget, len(chall_runs),
-                )
-            inc_run = inc_runs[0]
-            chall_run = chall_runs[0]
-            if inc_run.budget > chall_run.budget:
-                self.logger.debug('Incumbent evaluated on higher budget than challenger (%.4f > %.4f), '
-                                  'not changing the incumbent',
-                                  inc_run.budget, chall_run.budget)
-                new_incumbent = incumbent
-            elif inc_run.budget < chall_run.budget:
-                self.logger.debug('Challenger evaluated on higher budget than incumbent (%.4f > %.4f), '
-                                  'changing the incumbent',
-                                  chall_run.budget, inc_run.budget)
+            self.logger.debug("Incumbent (%.4f) is at least as good as the challenger (%.4f) on budget %.4f.",
+                              inc_cost, chall_cost, inc_run.budget)
+            if log_traj and self.stats.inc_changed == 0:
+                # adding incumbent entry
+                self.stats.inc_changed += 1  # first incumbent
+                self.traj_logger.add_entry(train_perf=inc_cost,
+                                           incumbent_id=self.stats.inc_changed,
+                                           incumbent=incumbent)
+            new_incumbent = incumbent
+
+        return new_incumbent
+
+    def _compare_configs_across_budgets(self,
+                                        challenger: Configuration,
+                                        incumbent: Configuration,
+                                        run_history: RunHistory,
+                                        log_traj: bool = True) -> typing.Optional[Configuration]:
+        """
+        compares challenger with current incumbent on any budget
+
+        Parameters
+        ----------
+        challenger : Configuration
+            promising configuration
+        incumbent : Configuration
+            best configuration so far
+        run_history : smac.runhistory.runhistory.RunHistory
+            stores all runs we ran so far
+        log_traj : bool
+            whether to log changes of incumbents in trajectory
+
+        Returns
+        -------
+        typing.Optional[Configuration]
+            incumbent configuration
+        """
+        # compare challenger and incumbent based on cost
+        chall_cost = min(run_history.get_all_costs_for_config(challenger))
```

Comment:
Do you think it would be possible to change the `runhistory.get_cost` so that it doesn't average the cost, but the min cost returned here? Or add a new function `runhistory.get_min_cost`? Maybe this would require introducing a new internal data structure `self._min_cost_per_config` which is used for this setting.